apiserver/facades/agent/secretsmanager/secrets.go

```diff
@@ -344,20 +344,93 @@ func (s *SecretsManagerAPI) GetSecretContentInfo(args params.GetSecretContentArg
 	return result, nil
 }
 
-func (s *SecretsManagerAPI) getSecretContent(arg params.GetSecretContentArg) (*secrets.ContentParams, error) {
-	uri, err := coresecrets.ParseURI(arg.URI)
+func (s *SecretsManagerAPI) getOwnerSecretMetadata(uri *coresecrets.URI, label string) (*coresecrets.SecretMetadata, error) {
+	if uri == nil {
+		var err error
+		uri, err = s.secretsBackend.GetURIBySecretLabel(label, s.authTag)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	md, err := s.secretsBackend.GetSecret(uri)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	if !s.canRead(uri, s.authTag) {
-		return nil, apiservererrors.ErrPerm
+	if s.authTag.String() != md.OwnerTag {
+		// Invisible for non owner users.
+		return nil, errors.NotFoundf("secret %s, label %q", uri, label)
+	}
+	return md, nil
+}
+
+func (s *SecretsManagerAPI) getSecretContent(arg params.GetSecretContentArg) (*secrets.ContentParams, error) {
+	// Only the owner can access secrets via the secret label (Note: the leader unit is not the owner of the application secrets).
+	// Non owners have to use secret URI or consumer label.
+
+	if arg.URI == "" && arg.Label == "" {
+		return nil, errors.NewNotValid(nil, "both uri and label are empty")
 	}
+
+	getSecretValue := func(uri *coresecrets.URI, revision int) (*secrets.ContentParams, error) {
+		if !s.canRead(uri, s.authTag) {
+			return nil, apiservererrors.ErrPerm
+		}
+
+		val, providerId, err := s.secretsBackend.GetSecretValue(uri, revision)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		return &secrets.ContentParams{SecretValue: val, ProviderId: providerId}, nil
+	}
+
+	var uri *coresecrets.URI
+	var err error
+
+	if arg.URI != "" {
+		uri, err = coresecrets.ParseURI(arg.URI)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	// arg.Label is the secret label for owner.
+	md, err := s.getOwnerSecretMetadata(uri, arg.Label)
+	if err != nil && !errors.Is(err, errors.NotFound) {
+		return nil, errors.Trace(err)
+	}
+	if md != nil {
+		// Owner can access the content directly.
+		if arg.Update {
+			// We should not have cmd flag knowledge here but we have to check here because jujuc cannot know it.
+			return nil, errors.NewNotValid(nil, "secret owner cannot use --update")
+		}
+		return getSecretValue(md.URI, md.LatestRevision)
+	}
+
+	// arg.Label is the consumer label for consumers.
+	possibleUpdateLabel := arg.Label != "" && uri != nil
+
+	if uri == nil {
+		var err error
+		uri, err = s.secretsConsumer.GetURIByConsumerLabel(arg.Label, s.authTag)
+		if errors.Is(err, errors.NotFound) {
+			return nil, errors.NotFoundf("consumer label %q", arg.Label)
+		}
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
```

Comment:
This is the only URI look up we need to do - owners will always pass in a URI. If the owner attempts to use --label and also gives a URI, that's an error.